pkg/provisioner/ironic/provision_test.go

```diff
@@ -66,7 +66,7 @@ func TestProvision(t *testing.T) {
 		},
 		{
 			name: "active state",
-			ironic: testserver.NewIronic(t).Ready().Node(nodes.Node{
+			ironic: testserver.NewIronic(t).WithDefaultResponses().Node(nodes.Node{
 				ProvisionState: string(nodes.Active),
 				UUID:           nodeUUID,
 			}),
```

Comment:
@dhellmann Please take a look at this change. Because I'm not sure if this is okay.  I had to make change in order to pass unit tests. Once I added error handling as you [suggested](https://github.com/metal3-io/baremetal-operator/pull/784#discussion_r592578762) , unit tests started failing as below
```
$ make unit

   server.go:249: ironic: Cannot find any default response for [PATCH] /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58
    server.go:72: ironic: PATCH /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58 -> 
    --- FAIL: TestProvision/active_state (0.00s)
        server.go:17: inspector: new server created
        server.go:134: inspector: adding response for [GET] /v1
        server.go:134: inspector: adding response for [GET] /v1/introspection/33ce8659-7400-4c68-9535-d10766f07a58
        server.go:67: inspector: endpoint: http://127.0.0.1:41231/v1/
        provision_test.go:121: 
            	Error Trace:	provision_test.go:121
            	Error:      	Received unexpected error:
            	            	EOF
            	            	failed to update ironic node automated_clean
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.(*ironicProvisioner).updateAutomatedClean
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/ironic.go:386
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.(*ironicProvisioner).Provision
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/ironic.go:1386
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.TestProvision.func1
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/provision_test.go:111
            	            	testing.tRunner
            	            		/usr/local/go/src/testing/testing.go:1194
            	            	runtime.goexit
            	            		/usr/local/go/src/runtime/asm_amd64.s:1371
            	            	could not update ironic node automated_clean field
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.(*ironicProvisioner).Provision
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/ironic.go:1387
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.TestProvision.func1
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/provision_test.go:111
            	            	testing.tRunner
            	            		/usr/local/go/src/testing/testing.go:1194
            	            	runtime.goexit
            	            		/usr/local/go/src/runtime/asm_amd64.s:1371
            	Test:       	TestProvision/active_state
```

```
$ make unit | grep PATCH
...
    server.go:72: ironic: PATCH /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58 -> {"automated_clean":null,"uuid":"33ce8659-7400-4c68-9535-d10766f07a58","name":"","power_state":"","target_power_state":"","provision_state":"","target_provision_state":"","maintenance":false,"maintenance_reason":"","fault":"","last_error":"","reservation":"","driver":"","driver_info":null,"driver_internal_info":null,"properties":null,"instance_info":null,"instance_uuid":"","chassis_uuid":"","extra":null,"console_enabled":false,"raid_config":null,"target_raid_config":null,"clean_step":null,"deploy_step":null,"resource_class":"","boot_interface":"","console_interface":"","deploy_interface":"","inspect_interface":"","management_interface":"","network_interface":"","power_interface":"","raid_interface":"","rescue_interface":"","storage_interface":"","traits":null,"vendor_interface":"","conductor_group":"","protected":false,"protected_reason":"","owner":""}
    server.go:249: ironic: Cannot find any default response for [PATCH] /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58
    server.go:72: ironic: PATCH /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58 -> 
    server.go:249: ironic: Cannot find any default response for [PATCH] /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58
    server.go:72: ironic: PATCH /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58 -> 
...
```

Replacing `testserver.NewIronic(t).Ready().Node` -> `testserver.NewIronic(t).WithDefaultResponses().Node` fixed the units tests. But I'm not 100% clear about those unit tests.

Reply:
I don't understand how my patch is breaking the existing unit tests.

Reply:
My guess: previously we would call PATCH only once for all operations. Now you've added another call, but there is no fixture for it.

Reply:
Do we actually have to do the PATCH in the Provision/Deprovision functions, or could we just add the update e.g to `getImageUpdateOptsForNode` ?

That would simplify the code changes, and reduce the test impact but perhaps I'm missing some corner cases which won't get handled in that case?

Reply:
When is it called? If it's only called before the deployment, it's too late to disable cleaning (it happens on the way from `manageable` to `available`).

Reply:
I'm not sure about `getImageUpdateOptsForNode`,  and the reason I used it in the Provision() & Deprovision() is because cleaning also can be enabled/disabled for provisioning & deprovisioning. I could not find any better place to call Ironic to update the field.

Reply:
@fmuyassarov in case of problems the mock's output has been designed to help you arrange the required calls in the test setup: the most frequent reason it's a change in the code that stimulates one (or more) API calls.
The usage of `WithDefaultResponses` could be handy when, in the test setup, you're not interested in the "usual" API calls, and so its goal it's to add a valid (mostly generic) answer for such calls. On the other side, it could hide some important interactions that instead could be worth to be mocked. If you remove `WithDefaultResponses` and run the test again, the output should provide enough info to detect which call requires to be mocked (ie by looking at the messages in the format `... Cannot find any default response for ...`).

Reply:
Thank you @andfasano for the explanation! That's what I got now...

```
       server.go:72: ironic: GET /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58 -> {"automated_clean":null,"uuid":"33ce8659-7400-4c68-9535-d10766f07a58","name":"","power_state":"","target_power_state":"","provision_state":"active","target_provision_state":"","maintenance":false,"maintenance_reason":"","fault":"","last_error":"","reservation":"","driver":"","driver_info":null,"driver_internal_info":null,"properties":null,"instance_info":null,"instance_uuid":"","chassis_uuid":"","extra":null,"console_enabled":false,"raid_config":null,"target_raid_config":null,"clean_step":null,"deploy_step":null,"resource_class":"","boot_interface":"","console_interface":"","deploy_interface":"","inspect_interface":"","management_interface":"","network_interface":"","power_interface":"","raid_interface":"","rescue_interface":"","storage_interface":"","traits":null,"vendor_interface":"","conductor_group":"","protected":false,"protected_reason":"","owner":""}
    server.go:236: ironic: found default response for /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58: {200, {"automated_clean":null,"uuid":"{id}","name":"","power_state":"","target_power_state":"","provision_state":"","target_provision_state":"","maintenance":false,"maintenance_reason":"","fault":"","last_error":"","reservation":"","driver":"","driver_info":null,"driver_internal_info":null,"properties":null,"instance_info":null,"instance_uuid":"","chassis_uuid":"","extra":null,"console_enabled":false,"raid_config":null,"target_raid_config":null,"clean_step":null,"deploy_step":null,"resource_class":"","boot_interface":"","console_interface":"","deploy_interface":"","inspect_interface":"","management_interface":"","network_interface":"","power_interface":"","raid_interface":"","rescue_interface":"","storage_interface":"","traits":null,"vendor_interface":"","conductor_group":"","protected":false,"protected_reason":"","owner":""}}
    server.go:72: ironic: PATCH /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58 -> {"automated_clean":null,"uuid":"33ce8659-7400-4c68-9535-d10766f07a58","name":"","power_state":"","target_power_state":"","provision_state":"","target_provision_state":"","maintenance":false,"maintenance_reason":"","fault":"","last_error":"","reservation":"","driver":"","driver_info":null,"driver_internal_info":null,"properties":null,"instance_info":null,"instance_uuid":"","chassis_uuid":"","extra":null,"console_enabled":false,"raid_config":null,"target_raid_config":null,"clean_step":null,"deploy_step":null,"resource_class":"","boot_interface":"","console_interface":"","deploy_interface":"","inspect_interface":"","management_interface":"","network_interface":"","power_interface":"","raid_interface":"","rescue_interface":"","storage_interface":"","traits":null,"vendor_interface":"","conductor_group":"","protected":false,"protected_reason":"","owner":""}
    server.go:67: ironic: endpoint: http://127.0.0.1:39187/v1/
    server.go:72: ironic: GET /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58 -> {"automated_clean":null,"uuid":"33ce8659-7400-4c68-9535-d10766f07a58","name":"","power_state":"","target_power_state":"","provision_state":"cleaning","target_provision_state":"","maintenance":false,"maintenance_reason":"","fault":"","last_error":"","reservation":"","driver":"","driver_info":null,"driver_internal_info":null,"properties":null,"instance_info":null,"instance_uuid":"","chassis_uuid":"","extra":null,"console_enabled":false,"raid_config":null,"target_raid_config":null,"clean_step":null,"deploy_step":null,"resource_class":"","boot_interface":"","console_interface":"","deploy_interface":"","inspect_interface":"","management_interface":"","network_interface":"","power_interface":"","raid_interface":"","rescue_interface":"","storage_interface":"","traits":null,"vendor_interface":"","conductor_group":"","protected":false,"protected_reason":"","owner":""}
    server.go:249: ironic: Cannot find any default response for [PATCH] /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58
    server.go:72: ironic: PATCH /v1/nodes/33ce8659-7400-4c68-9535-d10766f07a58 -> 
    --- FAIL: TestProvision/other_state:_Cleaning (0.00s)
        server.go:17: inspector: new server created
        server.go:134: inspector: adding response for [GET] /v1
        server.go:134: inspector: adding response for [GET] /v1/introspection/33ce8659-7400-4c68-9535-d10766f07a58
        server.go:67: inspector: endpoint: http://127.0.0.1:37975/v1/
        provision_test.go:113: 
            	Error Trace:	provision_test.go:113
            	Error:      	Not equal: 
            	            	expected: true
            	            	actual  : false
            	Test:       	TestProvision/other_state:_Cleaning
        provision_test.go:114: 
            	Error Trace:	provision_test.go:114
            	Error:      	Not equal: 
            	            	expected: 10s
            	            	actual  : 0s
            	Test:       	TestProvision/other_state:_Cleaning
        provision_test.go:121: 
            	Error Trace:	provision_test.go:121
            	Error:      	Received unexpected error:
            	            	EOF
            	            	failed to update ironic node automated_clean
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.(*ironicProvisioner).updateAutomatedClean
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/ironic.go:386
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.(*ironicProvisioner).Provision
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/ironic.go:1386
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.TestProvision.func1
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/provision_test.go:111
            	            	testing.tRunner
            	            		/usr/local/go/src/testing/testing.go:1194
            	            	runtime.goexit
            	            		/usr/local/go/src/runtime/asm_amd64.s:1371
            	            	could not update ironic node automated_clean field
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.(*ironicProvisioner).Provision
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/ironic.go:1387
            	            	github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic.TestProvision.func1
            	            		/home/fmuyassarov/go/src/github/baremetal-operator/pkg/provisioner/ironic/provision_test.go:111
            	            	testing.tRunner
            	            		/usr/local/go/src/testing/testing.go:1194
            	            	runtime.goexit
            	            		/usr/local/go/src/runtime/asm_amd64.s:1371
            	Test:       	TestProvision/other_state:_Cleaning
```

Reply:
To fix it, as also @dtantsur suggested, you need to instruct the Ironic mock to give a valid answer for a node PATCH using `NodeUpdate`, ie:

```
              {
			name: "active state",
			ironic: testserver.NewIronic(t).Ready().Node(nodes.Node{
				ProvisionState: string(nodes.Active),
				UUID:           nodeUUID,
			}).NodeUpdate(nodes.Node{
				UUID: nodeUUID,
			}),
			expectedRequestAfter: 0,
			expectedDirty:        false,
		},
```

Reply:
Great, that indeed fixed my issue! Thank you @andfasano ! Please take a look.